packages/stellar-axelar-std-derive/src/storage.rs

```diff
@@ -0,0 +1,176 @@
+}
+
+pub fn contractstorage(input: &DeriveInput) -> TokenStream {
+    let name = &input.ident;
+
+    let Data::Enum(DataEnum { variants, .. }) = &input.data else {
+        panic!("contractstorage can only be used on enums.");
+    };
+
+    let transformed_variants: Vec<_> = variants.iter().map(transform_variant).collect();
+
+    let storage_fns: Vec<_> = variants
+        .iter()
+        .map(|variant| {
+            let attrs = storage_attributes(&variant.attrs);
+            storage_fns(name, variant, &attrs)
+        })
+        .collect();
+
+    let output = quote! {
+        #[contracttype]
+        pub enum #name {
+            #(#transformed_variants,)*
+        }
+
+        impl #name {
+            #(#storage_fns)*
+        }
+    };
+
+    output
+}
+
+fn transform_variant(variant: &Variant) -> TokenStream {
+    let variant_name = &variant.ident;
+
+    match &variant.fields {
+        Fields::Unit => {
+            quote! {
+                #variant_name
+            }
+        }
+        Fields::Named(FieldsNamed { named, .. }) => {
+            let types = named.iter().map(|f| &f.ty);
+            quote! {
+                #variant_name(#(#types),*)
+            }
+        }
+        _ => panic!("Only unit variants or named fields are supported in storage enums."),
+    }
+}
+
+fn storage_attributes(attrs: &[Attribute]) -> StorageAttributes {
+    let mut storage_attributes = StorageAttributes::default();
+
+    for attr in attrs {
+        let path_str = attr.path().to_token_stream().to_string();
+
+        match path_str.as_str() {
+            "instance" => {
+                storage_attributes.storage_type = StorageType::Instance;
+            }
+            "persistent" => {
+                storage_attributes.storage_type = StorageType::Persistent;
+            }
+            "temporary" => {
+                storage_attributes.storage_type = StorageType::Temporary;
+            }
+            "value" => {
+                if let Meta::List(list) = &attr.meta {
+                    if let Ok(ty) = list.parse_args::<Type>() {
+                        storage_attributes.value_type = Some(ty);
+                    }
+                }
+            }
+            _ => {}
+        }
+    }
+
+    storage_attributes
+}
+
+fn storage_fns(
+    enum_name: &Ident,
+    variant: &Variant,
+    storage_attrs: &StorageAttributes,
+) -> TokenStream {
+    let variant_ident = &variant.ident;
+
+    let (field_names, field_types) = match &variant.fields {
+        Fields::Unit => (vec![], vec![]),
+        Fields::Named(fields) => {
+            let names = fields.named.iter().map(|f| &f.ident).collect();
+            let types = fields.named.iter().map(|f| &f.ty).collect();
+            (names, types)
+        }
+        _ => panic!("Only unit variants or named fields are supported in storage enums."),
+    };
+
+    let value_type = storage_attrs
+        .value_type
+        .as_ref()
+        .expect("value type required");
```

Comment:
Bad pattern. The StorageAttributes type shouldn't allow an invalid value to begin with. In Rust, types enforce safety / correctness, so they should be defined in a way that prevents invalid values/combinations altogether